torch/distributed/_shard/checkpoint/utils.py

```diff
@@ -0,0 +1,236 @@
+        """
+        object_list = [object]
+        if self.use_dist:
+            dist.broadcast_object_list(
+                object_list=object_list,
+                group=self.group,
+                src=self.coordinator_rank)
+        return cast(T, object_list[0])
+
+    def gather_object(self, object: T) -> Optional[List[T]]:
+        """
+        Same as c10d::gather_object but works without distributed enabled.
+        """
+        if self.use_dist:
+            gather_objs = cast(List[T], [None] * dist.get_world_size(self.group)) if self.is_coordinator else None
+
+            dist.gather_object(
+                obj=object,
+                object_gather_list=gather_objs if self.is_coordinator else None,
+                dst=self.coordinator_rank,
+                group=self.group
+            )
+            # technically the type is shou
+            result = gather_objs
+        else:
+            result = [object]
+        return result
+
+    def all_gather_object(self, object: T) -> List[T]:
+        """
+        Same as c10d::all_gather_object but works without distributed enabled.
+        """
+        if self.use_dist:
+            gather_objs = cast(List[T], [None] * dist.get_world_size(self.group))
+
+            dist.all_gather_object(
+                object_list=gather_objs,
+                obj=object,
+                group=self.group
+            )
+        else:
+            gather_objs = [object]
+        return gather_objs
+
+    def scatter_object(self, object_list: Optional[List[T]]) -> T:
+        """
+        Same as c10d::scatter_object but works without distributed enabled.
+        """
+        if self.use_dist:
+            gather_result = cast(List[T], [None])
+            dist.scatter_object_list(
+                scatter_object_output_list=gather_result,
+                scatter_object_input_list=object_list if self.is_coordinator else None,
+                src=self.coordinator_rank,
+                group=self.group
+            )
+
+            local_reply = gather_result[0]
+        else:
+            assert object_list is not None
+            local_reply = object_list[0]
+        return local_reply
+
+    def reduce_scatter(
+        self,
+        step: str,
+        map_fun: Callable[[], T],
+        reduce_fun: Callable[[List[T]], List[R]]
+    ) -> R:
+        """
+        Compute a value on each rank, then do centralized reduce on a single rank, followed by a scatter.
+
+        This method operates in the following way:
+            Run ``map_cp`` on all ranks
+            Gather results on rank 0
+            Call ``reduce_cb`` on all those values
+            Scatter to each rank part of the result.
+        """
+        local_data: Union[BaseException, T]
+        try:
+            local_data = map_fun()
+        except BaseException as e:
+            local_data = e
+
+        all_data = self.gather_object(local_data)
```

Comment:
We have to reimplement `reduce_scatter` because we're reducing by an user function which is not supported by c10d.

In addition to that, we have to handle the case where `map_fun` raises an exception, which requires a different form of reduction.

Reply:
All `_object` functions are wrappers around c10d collectives that handle distributed not being initialized.